### PR TITLE
Make amax reduction optional

### PR DESCRIPTION
**Pull request by ksivaman**
Allow possibility to opt out of global amax reduction via the FP8 recipe

Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

### CODE REVIEW

File: transformer_engine/common/recipe.py [hunk 1]
```diff
@@ -98,6 +98,9 @@ def scaling_factor_compute(amax: Tensor,
     override_linear_precision: Tuple(bool, bool, bool), default=(False, False, False)
                               Whether or not the execute the `fprop`, `dgrad`, and `wgrad`
                               GEMMs (respectively) in higher precision when using FP8.
+    reduce_amax: bool, default = `True`
+                If set to `True`, the `amax` value for fp8 tensors is reduced across the given
+                `fp8_group` using the `fp8_autocast` API after every iteration.
```
Comment (ptrendx) [inline transformer_engine/common/recipe.py:103]: This is unclear to me - what is `fp8_autocast`'s role in this sentence?

Comment (ksivaman) [reply transformer_engine/common/recipe.py:103]: It is to mention that the global amax reduction happens across the `fp8_group` argument, which is specified via the `fp8_autocast` API

File: transformer_engine/pytorch/module.py [hunk 2]
```diff
@@ -368,7 +373,7 @@ def post_backward(
         if not fp8:
             return
 
-        if fp8_meta["first_module"]:
+        if fp8_meta["first_module"] and fp8_meta["recipe"].reduce_amax:
```
[the remark this answers is not in the document]
Comment (ksivaman) [reply transformer_engine/pytorch/module.py:376]: Yes, this call should be skipped. However, there are also some other amax reduction setup procedures in `pre_forward` that we can skip, which I will add.

File: transformer_engine/common/recipe.py [hunk 3]
```diff
@@ -98,6 +98,14 @@ def scaling_factor_compute(amax: Tensor,
...
                               Whether or not the execute the `fprop`, `dgrad`, and `wgrad`
                               GEMMs (respectively) in higher precision when using FP8.
+    reduce_amax: bool, default = `True`
+                By default, if `torch.distributed` is initialized, the `amax` value for fp8
+                tensors is reduced across the specified `fp8_group` using the `fp8_autocast`
+                API after every iteration. This keeps the amaxes and scaling factors synced
+                across the given distributed group. If set to `False`, this reduction is
+                skipped and every GPU maintains localized amax and scaling factors. To ensure
+                results are numerically identical across checkpointing boundaries, all ranks
+                must checkpoint in order to store the local tensors.
```
Comment (ptrendx) [inline transformer_engine/common/recipe.py:108]: I still don't quite like this part about fp8_autocast. Maybe something like
```
the amax value for FP8 tensors is reduced across the fp8_group (specified in the fp8_autocast call).
```

File: transformer_engine/common/recipe.py [hunk 4]
```diff
@@ -98,6 +98,14 @@ def scaling_factor_compute(amax: Tensor,
     override_linear_precision: Tuple(bool, bool, bool), default=(False, False, False)
                               Whether or not the execute the `fprop`, `dgrad`, and `wgrad`
                               GEMMs (respectively) in higher precision when using FP8.
+    reduce_amax: bool, default = `True`
+                By default, if `torch.distributed` is initialized, the `amax` value for fp8
+                tensors is reduced across the specified `fp8_group` using the `fp8_autocast`
+                API after every iteration. This keeps the amaxes and scaling factors synced
+                across the given distributed group. If set to `False`, this reduction is
+                skipped and every GPU maintains localized amax and scaling factors. To ensure
```
Comment (ptrendx) [inline transformer_engine/common/recipe.py:106]: `local` instead of `localized`?

Comment (ptrendx) [reply transformer_engine/common/recipe.py:106]: Also, we should make it more clear that the following sentence is only about the `False` setting. Maybe something like this:
```
To ensure results are numerically identical across checkpointing boundaries in this case,
```
?

File: transformer_engine/pytorch/module.py [hunk 5]
```diff
@@ -180,10 +184,13 @@ def set_extra_state(self, state: Union[List[Any], None]) -> None:
         # Restore global FP8 buffer state.
         set_global_fp8_buffer(state[4])
         self.fp8_meta["update_amax_and_scale_fwd"] = state[5]
-        self.fp8_meta["global_fp8_buffer_pos_fwd"] = state[6]
-        self.fp8_meta["global_fp8_buffer_pos_bwd"] = state[7]
-        self.fp8_meta["autocast_id_fwd"] = state[8]
-        self.fp8_meta["autocast_id_bwd"] = state[9]
+
+        # Load extra items for amax reduction
+        if len(state) > 6:
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:189]: This seems brittle :-(.

Comment (ptrendx) [reply transformer_engine/pytorch/module.py:189]: Does state need to be a list?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:189]: Anything pickleable should work

File: transformer_engine/common/recipe.py [hunk 6]
```diff
@@ -98,6 +98,14 @@ def scaling_factor_compute(amax: Tensor,
     override_linear_precision: Tuple(bool, bool, bool), default=(False, False, False)
                               Whether or not the execute the `fprop`, `dgrad`, and `wgrad`
                               GEMMs (respectively) in higher precision when using FP8.
+    reduce_amax: bool, default = `True`
+                By default, if `torch.distributed` is initialized, the `amax` value for FP8
+                tensors is reduced across the `fp8_group` (specified in the `fp8_autocast`
+                call). API after every iteration. This keeps the amaxes and scaling factors
```
Comment (ptrendx) [inline transformer_engine/common/recipe.py:104]: "API after every iteration." What?

Comment (ksivaman) [reply transformer_engine/common/recipe.py]: Fixed

File: transformer_engine/pytorch/module.py [hunk 7]
```diff
@@ -300,22 +303,25 @@ def pre_forward(self, inp: torch.Tensor, num_gemms: int = 1) -> None:
...
+                set_amax_buffer_key_deletion(self.fp8_meta, forward=True)
             else:
-                self.fp8_meta["autocast_id_fwd"] = get_fp8_context_id()
+                amax_and_scale_update(self.fp8_meta, True)
 
-            add_amax_to_global_buffer(self.fp8_meta, forward=True)
+        if self.fp8 and torch.is_grad_enabled() and self.training:
+            # Setup for amax reduction
+            if self.fp8_meta["recipe"].reduce_amax:
+                self.fp8_meta["first_module"] = is_first_fp8_module()
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:318]: Why is this part skipped when not reducing amax? You still need to update the scaling factor in that case at the end of fwd and backward so I think the data on whether it is the first fp8 module should still be updated?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:318]: Scaling factor and amax updates are local per module and only the reductions are aggregated. So information about first module isn't needed for this case.

Comment (ptrendx) [reply transformer_engine/pytorch/module.py:318]: But how can you update the scaling factors if you don't have amax yet? If the amax reduction is aggregated then the scaling factors update also has to be, right?

Comment (ksivaman) [reply transformer_engine/pytorch/module.py:318]: The aggregation of the amaxes for reduction is a separate step from updating the amax and scale factors. The latter is done at the beginning of the following iteration and is localized per module (not aggregated).

**Review by ptrendx — APPROVED** [review]
Please mark loading of the old checkpoint as deprecated, otherwise LGTM.

File: transformer_engine/pytorch/module.py [hunk 8]
```diff
@@ -169,6 +172,11 @@ def set_extra_state(self, state: Union[List[Any], None]) -> None:
 
         # Maintain backward compatibility with v0.2.0 and older.
         if isinstance(state, list):
+            warnings.warn(
+                "This format of checkpointing deprecated and will be"
+                "removed in future releases of TransformerEngine"
```
Comment (ptrendx) [inline transformer_engine/pytorch/module.py:177]:
```suggestion
                "This checkpoint format is deprecated and will be"
                "removed in a future release of Transformer Engine"
```

**Pull request merged**